### PR TITLE
[COMPLETE] Sort XML according to provided schema(s)

### LINKED ISSUE #146: Sort outputted XML according to an XML schema
This is necessary for it to validate against a schema that uses sequences, such as the IATI one https://github.com/IATI/IATI-Schemas/blob/version-2.02/iati-activities-schema.xsd

### PR DESCRIPTION
**Pull request by andylolz**
Finishes off work on [the ODS 146-sort-xml branch](https://github.com/OpenDataServices/flatten-tool/tree/146-sort-xml).

~~Refs~~ Fixes #146.

~~Note that this adds multiple copies of the IATI v2.03 schemas to this repo, which is why this PR is huge. It might be desirable not to do this (e.g. use a simplified non-IATI schema in either the tests or the docs or both).~~ <-- fixed in d5a8bcd.

### CODE REVIEW

**Comment by andylolz** [comment]
It seemed like [ODS 146-sort-xml](https://github.com/OpenDataServices/flatten-tool/tree/146-sort-xml) was taking the route of hardcoding IATI XML. I’ve instead added means to pass in the XML schemas as a separate parameter (to avoid adding more coupling between flatten-tool and IATI.)

**Comment by andylolz** [comment]
I can send the corresponding fix for CoVE, in case that’s useful.

**Comment by andylolz** [comment]
@Bjwebb great to see you’re taking a look at this. Given you’ve added commits, do you want me to review?

**Comment by Bjwebb** [comment]
And here's the related change to CoVE https://github.com/OpenDataServices/cove/compare/flattentool-146-sort-xml (will be updated once this PR is merged).

File: flattentool/xml_output.py [hunk 1]
```diff
@@ -70,16 +71,19 @@ def dict_to_xml(data, tagname, toplevel=True, nsmap=None):
...
+    if xml_schemas is not None:
+        schema_dict = IATISchemaWalker(xml_schemas).create_schema_dict(root_list_path)
+        for element in root:
+            sort_iati_element(element, schema_dict)
     comment = ET.Comment('XML generated by flatten-tool')
     root.insert(0, comment)
     if USING_LXML:
         return ET.tostring(root, pretty_print=True)
     else:
-        indent(root)
```
Comment (andylolz) [inline flattentool/xml_output.py]: This fixes an error on master, but it’s unrelated to this PR.

File: .travis.yml [hunk 2]
```diff
@@ -2,9 +2,9 @@ sudo: false
 language: python
 python:
   - "2.7"
-  - "3.3"
```
Comment (andylolz) [inline .travis.yml]: Seems legit, but also seems unrelated to this PR.

File: .travis.yml [hunk 3]
```diff
@@ -2,9 +2,9 @@ sudo: false
...
   - "3.4"
   - "3.5"
+  - "3.6"
```
Comment (andylolz) [inline .travis.yml]: Same as https://github.com/OpenDataServices/flatten-tool/pull/196/files#r207384510

File: flattentool/sort_iati.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    for child in sorted(children, key=index_key):
+        element.append(child)
+        sort_iati_element(child, schema_subdict.get(child.tag, {}))
+
+
+def sort_iati_xml_file(input_file, output_file, schemas):
+    """
+    Sort an IATI XML file according to the schema.
+    """
+    schema_dict = IATISchemaWalker(schemas).create_schema_dict('iati-activity')
```
Comment (andylolz) [inline flattentool/sort_iati.py]: Elsewhere, iati-activity is generalised away as `root_list_path`. So it seems a bit untidy here.

I actually wonder about removing `sort_iati_xml_file` onwards completely, since it’s pretty useless anyway.

**Comment by andylolz** [comment]
Looks good to me!

I’d just suggest removing a chunk of code from the end of sort_iati.py, because it’s basically cruft anyway, but also because it’s one of the few remaining hardcoded element names so it just looks a bit messy.

Aside from that – lovely!

File: flattentool/sort_iati.py (hunk 4 above)
Comment (andylolz) [inline flattentool/sort_iati.py]: Elsewhere, iati-activity is generalised away as `root_list_path`. So it seems a bit untidy here.

I’d actually suggest removing `sort_iati_xml_file` onwards completely, since it’s pretty useless anyway.

**Comment by andylolz** [comment]
> Sorry this has taken a while. We've not had any IATI sprints, so this has been slowly working its way up my misc TODO list.

No worries! I’ve just realised I only picked this one up because @robredpath said it was difficult!

File: .travis.yml (hunk 3 above)
Comment (Bjwebb) [reply .travis.yml]: This change is now on the master branch, so doesn't appear in the diff.

File: flattentool/sort_iati.py (hunk 4 above)
Comment (Bjwebb) [reply flattentool/sort_iati.py]: Plus, this still exists at https://github.com/OpenDataServices/iati-utils/blob/master/sort_iati.py

Thanks for the extra commit, I'll also tweak the header now its not also a standalone command.

File: CHANGELOG.md [hunk 5]
```diff
@@ -6,6 +6,16 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
 ## [Unreleased]
```
Comment (Bjwebb) [inline CHANGELOG.md]: @odscjames I'm planning on making a flatten-tool release on Monday, so that I can get the changes into CoVE.  The CHANGELOG will be updated when this PR is merged. Just thought I'd check that this looks good to you.

File: flattentool/sort_iati.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+# Namespaces necessary for opening schema files
+namespaces = {
+    'xsd': 'http://www.w3.org/2001/XMLSchema'
+}
+
+
+class IATISchemaWalker(object):
+    """
+    Class for converting an IATI XML schema to documentation in the
+    reStructuredText format.
```
Comment (kindly) [inline flattentool/sort_iati.py]: Would be good to update the doc-string to document what it does do now.

Comment (andylolz) [reply flattentool/sort_iati.py]: Good point! But it’s a slippery slope…!

Comment (andylolz) [reply flattentool/sort_iati.py]: Done in 7322da3… but that led me to fixing a load of other stuff.

File: flattentool/xml_output.py (hunk 1 above)
Comment (odscjames) [reply flattentool/xml_output.py]: Can there be a note of what error was fixed in the changelog? Thanks

File: CHANGELOG.md (hunk 5 above)
Comment (odscjames) [reply CHANGELOG.md]: Sounds good, but remember there is also the version number in setup.py to bump. 

I guess in this case you know nothing else will go in by then (so there will be no merge issues), but this might be something to think about in the future.

File: flattentool/xml_output.py (hunk 1 above)
Comment (andylolz) [reply flattentool/xml_output.py]: Ah, 4513cd34 introduced an error which this fixes.

But it’s only relevant when using stdlib etree which is unsupported anyway, so maybe not worth mentioning?

Comment (andylolz) [reply flattentool/xml_output.py]: Ah, 4513cd34 introduced a bug which this fixes. The bugfix means using stdlib etree ought to work… But that’s an unsupported feature anyway, so maybe not worth mentioning?

(Sidenote: personally, I’d suggest either supporting stuff or getting rid of it. I’m not terribly keen on making stuff sort of work, but with warnings saying it’s unsupported. This is illustrative of that. The unsupported feature didn’t work due to a bug; this fixes the bug so now the unsupported feature maybe works. Should that go in the changelog? It’s still unsupported, so maybe not…?)

Comment (odscjames) [reply flattentool/xml_output.py]: I'm thinking of a change log as being useful for 2 things (tho there may be more):

  *  as a user of the software, i am changing from using an older version to the later version and I want to know what has changed so I can make sure my use still works.
  *  as a developer working on the software, i want to be able to work out why the code that is there does what it does (git commit messages should help a bit with this, but more info is good)

None of that depends on whether the feature is supported or not; in either case I still want to know what has changed. So on first think (so to speak) I'm gonna say I think it should be in the changelog. (It can be marked unsupported.)

In the long term, I think I agree with you on either supporting something properly - and making sure it's clearly documented so that new developers coming in have a chance of working out what is going on - or just removing it. Otherwise we have this weird situation where a new developer might do some work which is great with all the supported functions but then they are told "P.R. refused, this breaks a thing you had no chance of knowing about, you fool!". (Not that we tell people they are fools but you can see how this would be frustrating.) Basically, it increases barriers to knowledge sharing.

Comment (andylolz) [reply flattentool/xml_output.py]: kk – I’ve sent this as a separate PR (#212) and added a changelog entry there.

File: flattentool/xml_output.py [hunk 7]
```diff
@@ -13,7 +13,7 @@
     USING_LXML = False
     warn('Using stdlib etree may work, but is not supported. Please install lxml.')
 from flattentool.exceptions import DataErrorWarning
-from flattentool.sort_iati import sort_element, XMLSchemaWalker
```
Comment (Bjwebb) [inline flattentool/xml_output.py]: Thanks for this rename! I was thinking the same thing.

Comment (andylolz) [reply flattentool/xml_output.py]: oh, cool! I generally don’t do this sort of thing in external PRs, because renaming things is often a bit controversial!

**Comment by andylolz** [comment]
:rocket: :balloon: :star2: :balloon: :star2: :balloon: :star2: :balloon: :rocket:

Thanks all!
